internal/core/src/exec/expression/Expr.h

```diff
@@ -0,0 +1,321 @@
+            is_index_mode_ ? current_index_chunk_pos_ : current_data_chunk_pos_;
+        auto current_rows = current_chunk * size_per_chunk_ + current_chunk_pos;
+        return current_rows + batch_size_ >= num_rows_
+                   ? num_rows_ - current_rows
+                   : batch_size_;
+    }
+
+    template <typename T, typename FUNC, typename... ValTypes>
+    int64_t
+    ProcessDataChunks(
+        FUNC func,
+        std::function<bool(const milvus::SkipIndex&, FieldId, int)> skip_func,
+        bool* res,
+        ValTypes... values) {
+        int64_t processed_size = 0;
+
+        for (size_t i = current_data_chunk_; i < num_data_chunk_; i++) {
+            auto data_pos =
+                (i == current_data_chunk_) ? current_data_chunk_pos_ : 0;
+            auto size = (i == (num_data_chunk_ - 1))
+                            ? (segment_->type() == SegmentType::Growing
+                                   ? num_rows_ % size_per_chunk_ - data_pos
+                                   : num_rows_ - data_pos)
+                            : size_per_chunk_ - data_pos;
+
+            size = std::min(size, batch_size_ - processed_size);
+
+            auto& skip_index = segment_->GetSkipIndex();
+            if (!skip_func || !skip_func(skip_index, field_id_, i)) {
+                auto chunk = segment_->chunk_data<T>(field_id_, i);
+                const T* data = chunk.data() + data_pos;
+                func(data, size, res + processed_size, values...);
+            }
+
+            processed_size += size;
+            if (processed_size >= batch_size_) {
+                current_data_chunk_ = i;
+                current_data_chunk_pos_ = data_pos + size;
+                break;
+            }
+        }
+
+        return processed_size;
+    }
+
+    int
+    ProcessIndexOneChunk(FixedVector<bool>& result,
+                         size_t chunk_id,
+                         const FixedVector<bool>& chunk_res,
+                         int processed_rows) {
+        auto data_pos =
+            chunk_id == current_index_chunk_ ? current_index_chunk_pos_ : 0;
+        auto size = std::min(
+            std::min(size_per_chunk_ - data_pos, batch_size_ - processed_rows),
+            int64_t(chunk_res.size()));
+
+        result.insert(result.end(),
+                      chunk_res.begin() + data_pos,
+                      chunk_res.begin() + data_pos + size);
+        return size;
+    }
+
+    template <typename T, typename FUNC, typename... ValTypes>
+    FixedVector<bool>
+    ProcessIndexChunks(FUNC func, ValTypes... values) {
+        typedef std::
+            conditional_t<std::is_same_v<T, std::string_view>, std::string, T>
+                IndexInnerType;
+        using Index = index::ScalarIndex<IndexInnerType>;
+        FixedVector<bool> result;
+        int processed_rows = 0;
+
+        for (size_t i = current_index_chunk_; i < num_index_chunk_; i++) {
+            if (cached_index_chunk_id_ != i) {
```

Comment:
give some explain about cached_index_chunk_id_ and why this is necessary

Reply:
add comments done